src/test/java/com/meta/chatbridge/message/FBMessageHandlerTest.java

```diff
@@ -0,0 +1,368 @@
+                r -> createMessageRequest("{\"object\": \"page\"}", r),
+                false),
+            new TestArgument(
+                "valid body with no entries/messages",
+                200,
+                r -> createMessageRequest("{\"object\": \"page\", \"entry\": []}", r),
+                false),
+            new TestArgument(
+                "unknown entry objects",
+                200,
+                r -> createMessageRequest("{\"object\": \"page\", \"entry\": [{}]}", r),
+                false),
+            new TestArgument(
+                "valid body without any messaging objects",
+                200,
+                r ->
+                    createMessageRequest(
+                        "{\"object\": \"page\", \"entry\": [{\"messaging\": []}]}", r),
+                false),
+            new TestArgument(
+                "valid body with duplicate message",
+                200,
+                r -> createMessageRequest(MAPPER.writeValueAsString(duplicateMessage), r),
+                true),
+            new TestArgument(
+                "valid body with duplicate entry",
+                200,
+                r -> createMessageRequest(MAPPER.writeValueAsString(duplicateEntry), r),
+                true),
+            new TestArgument(
+                "malformed messaging object",
+                400,
+                r ->
+                    createMessageRequest(
+                        "{\"object\": \"page\", \"entry\": [{\"messaging\": [{}]}]}", r),
+                false),
+            new TestArgument(
+                "messaging object with no messages",
+                200,
+                r ->
+                    createMessageRequest(
+                        "{\"object\": \"page\", \"entry\": [{\"messaging\": [{\"recipient\": {\"id\": 123}, \"sender\": {\"id\": 123}, \"timestamp\": 0}]}]}",
+                        r),
+                false),
+            new TestArgument(
+                "invalid message object",
+                400,
+                r ->
+                    createMessageRequest(
+                        "{\"object\": \"page\", \"entry\": [{\"messaging\": [{\"recipient\": {\"id\": 123}, \"sender\": {\"id\": 123}, \"timestamp\": 0, \"message\": {}}]}]}",
+                        r),
+                false),
+            new TestArgument(
+                "missing hmac", 400, r -> createMessageRequest(SAMPLE_MESSAGE, r, false), false),
+            new TestArgument(
+                "invalid json", 400, r -> createMessageRequest("invalid_json.", r), false),
+            new TestArgument(
+                "valid json, invalid body", 400, r -> createMessageRequest("{}", r), false),
+            new TestArgument(
+                "invalid hmac",
+                400,
+                r ->
+                    createMessageRequest(SAMPLE_MESSAGE, r, false)
+                        .addHeader("X-Hub-Signature-256", "abcdef0123456789"),
+                false));
```

Comment:
Maybe harder to test but expired auth token for send? Do we want to throw a specific error for that to flag they need to update it?

And I think we have a 24 hour window to respond, maybe expired timestamp in case there's some delays?

Reply:
>Do we want to throw a specific error for that to flag they need to update it?
so, we actually require the app secret to validate the payload, so i want to put in a flow where we actually manage the token and it's lifecycle.

> And I think we have a 24 hour window to respond, maybe expired timestamp in case there's some delays?
great call out